functorch/experimental/output_dtype.py

```diff
@@ -0,0 +1,136 @@
+from functools import partial 
+
+import torch
+import torch.utils._pytree as pytree
+from torch.fx.experimental.proxy_tensor import (
+    ProxyTorchDispatchMode,
+    disable_proxy_modes_tracing,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+    unwrap_proxy,
+)
+from torch.utils._python_dispatch import (
+    _get_current_dispatch_mode,
+    _pop_mode_temporarily,
+)
+from torch._C import DispatchKey
+from torch._functorch.eager_transforms import (
+    _unwrap_all_tensors_from_functional,
+    _wrap_all_tensors_to_functional,
+)
+from torch._ops import HigherOrderOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+
+
+output_dtype = HigherOrderOperator("output_dtype")
+output_dtype.fallthrough(DispatchKey.PythonDispatcher)
+output_dtype.fallthrough(DispatchKey.PythonTLSSnapshot)
+output_dtype.fallthrough(DispatchKey.ADInplaceOrView)
+output_dtype.fallthrough(DispatchKey.BackendSelect)
+output_dtype.fallthrough(DispatchKey.AutocastCPU)
+
+
+def trace_output_dtype(proxy_mode, func_overload, op, out_dtype, *args):
+    if not isinstance(op, torch._ops.OpOverload):
+        raise ValueError("output_dtype's first argument must be an OpOverload")
+
+    with disable_proxy_modes_tracing():
+        casted_args = pytree.tree_map_only(
+            torch.Tensor, lambda arg: arg.to(dtype=out_dtype), args
+        )
+        out = op(*casted_args)
+
+    node_args = (op, out_dtype, *args)
+    proxy_args = pytree.tree_map(partial(unwrap_proxy, proxy_mode), node_args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function", func_overload, proxy_args, {}, name="output_dtype"
+    )
+    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
+
+
+@output_dtype.py_impl(DispatchKey.CompositeExplicitAutograd)
+def output_dtype_dense(
+    op: torch._ops.OpOverload,
+    out_dtype: torch.dtype,
+    *args
+):
+    casted_args = pytree.tree_map_only(
+        torch.Tensor, lambda arg: arg.to(dtype=out_dtype), args
+    )
+    res = op(*casted_args)
+    return res
+
+
+@output_dtype.py_impl(DispatchKey.Autograd)
+def output_dtype_autograd(
```

Comment:
Yeah so this is exactly what I would like to talk about. Take following example:
```
class M(torch.nn.Module):
  def __init__(self):
    self.linear1 = ...
    self.linear2 = ...

  def forward(self, x):
    return (self.linear2(self.linear1(x))
```

If I want to quantize M via quantization aware training, the current workflow on top of exported graph, will modify forward graph to effectively do
```
call this FakeQuantizeRepresentation
  def forward(self, x):
    x = fake_quantize(x, <some-quant-params-that-are-being-learnt-for-linear1>) # this step, effectively, just uses quant params to quantize and dequntize x
    x = self.linear1(x)
    x = fake_quantize(x, <some-quant-params-that-are-being-learnt-for-linear2>)
    x = self.linear2(x)
    return x
```

Above may not be entirely accurate, but close enough (@jerryzh168 @andrewor14 can confirm)

Fake quantize do not accurately capture numerical behavior of target hardware. While achieving bit-exact behavior might be hard, it will be very useful to do better approximation of numerics than what is done by fake quantize nodes. If we were able to do better approximation, then, the hypothesis is that, we will have computed loss that is more accurate, which may result in gradient updates that account for achievable accuracy on target hw. So what should `forward` look like with better numerics? We approximate that with integer compute e.g.

```
call this IntRepresentation
  def forward(self, x):
    x = quantize_int8(x, <some-quant-params-that-are-being-learnt-for-linear1>)
    x = self.int8_linear1(x, <some-quant-params-that-are-being-learnt-for-linear1>) # decomposition of this op uses "mixed_dtype" op introduces in this diff. Note that int8_linear will return int8 tensor
    x = self.int8_linear2(x, <some-quant-params-that-are-being-learnt-for-linear2>)
    return dequantize(x, <some-quant-params-that-are-being-learnt-for-linear2>)
```

Note that in IntRepresentation linear1 and linear2 are using int8 compute as opposed to FakeQuantizeRepresentation. If we can figure out autograd story for IntRepresentation that would be great. One of the challenges, and I have heard this from other vendors, is that int tensors dont store graidents.

If some of this is not super clear, I will draft something with more details. cc: @ezyang

Reply:
@kimishpatel thanks for writing down the details, did you get any feedback from hardware vendors how to do backward on integer compute? since it's not enough just to swap in the integer representation, we also need the computation to be differentiable in order to train.

Reply:
> feedback from hardware vendors how to do backward on integer compute?

They can do that today and some of them do, blockers for them is inability of int tensors to store gradient. But if we can enable this natively for decompositions that expressed via aten ops, that would add a lot of value.

> maybe one thing is to just use the fp version of the int8_linear operator, i.e. we convert all inputs to fp32

If we do this it will defeat the purpose of capturing compute with integer arithmetic. This is why it is somewhat non-trivial. Let me start a chat thread on this.

Reply:
> blockers for them is inability of int tensors to store gradient

is this in pytorch? I remember tried this once and it just says int Tensor doesn't have gradient, is it what you are referring to?